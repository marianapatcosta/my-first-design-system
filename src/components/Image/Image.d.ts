import React, { CSSProperties, KeyboardEventHandler, MouseEventHandler } from 'react';
export interface ImageProps {
    /**
     * element id
     */
    id?: string;
    /**
     * additional css class
     */
    className?: string;
    /**
     * additional style
     */
    style?: CSSProperties;
    /**
     * additional css class for image placeholder
     */
    placeholderClass?: string;
    /**
    /**
     * Image src attribute
     */
    src: string;
    /**
     * Image alt attribute
     */
    alt?: string;
    /**
     * is Image increasing size on click?
     */
    isZoomable?: boolean;
    /**
     * an event should be triggered on Image click? (Only available if isZoomable = false)
     */
    isClickable?: boolean;
    /**
     * Click handler (Only available if isZoomable = false)
     */
    onClick?: MouseEventHandler<HTMLImageElement> | KeyboardEventHandler<HTMLImageElement>;
}
export declare const Image: React.FC<ImageProps>;
