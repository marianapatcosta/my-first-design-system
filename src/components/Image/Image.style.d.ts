export declare const StyledImageOverlay: import("styled-components").StyledComponent<"div", any, {
    isZoomedIn?: boolean | undefined;
}, never>;
export declare const StyledImage: import("styled-components").StyledComponent<"img", any, {
    isClickable?: boolean | undefined;
    isLoaded?: boolean | undefined;
    isZoomedIn?: boolean | undefined;
    isZoomable?: boolean | undefined;
    wasImageClicked?: boolean | undefined;
}, never>;
export declare const StyledImagePlaceholder: import("styled-components").StyledComponent<"div", any, {}, never>;
