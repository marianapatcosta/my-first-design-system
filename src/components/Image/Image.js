var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { useEffect, useState, } from 'react';
import { isEventValid } from '../../utils/utils';
import { StyledImageOverlay, StyledImage, StyledImagePlaceholder, } from './Image.style';
export var Image = function (_a) {
    var placeholderClass = _a.placeholderClass, isClickable = _a.isClickable, isZoomable = _a.isZoomable, className = _a.className, onClick = _a.onClick, props = __rest(_a, ["placeholderClass", "isClickable", "isZoomable", "className", "onClick"]);
    var _b = useState(false), isLoaded = _b[0], setIsLoaded = _b[1];
    var _c = useState(false), isZoomedIn = _c[0], setIsZoomedIn = _c[1];
    var _d = useState(false), wasImageClicked = _d[0], setWasImageClicked = _d[1];
    var handleImageClick = function (event /* MouseEvent<HTMLImageElement> | KeyboardEvent<HTMLImageElement> */) {
        if (!isEventValid(event) || (!isClickable && !isZoomable))
            return;
        if (isZoomable) {
            setIsZoomedIn(function (prevIsZoomedIn) { return !prevIsZoomedIn; });
            !wasImageClicked && setWasImageClicked(true);
            return;
        }
        onClick && onClick(event);
    };
    useEffect(function () {
        if (isZoomedIn) {
            // document.getElementsByTagName('body')[0].style.overflow = 'hidden';
            document.documentElement.style.overflow = 'hidden';
            return;
        }
        // document.getElementsByTagName('body')[0].style.overflow = 'scroll';
        document.documentElement.style.overflow = 'scroll';
    }, [isZoomedIn]);
    return isZoomable ? (_jsxs(StyledImageOverlay, __assign({ isZoomedIn: isZoomedIn }, { children: [!isLoaded && (_jsx(StyledImagePlaceholder, { "data-testid": 'image-placeholder', className: placeholderClass || className }, void 0)),
            _jsx(StyledImage, __assign({ "aria-label": "click to " + (!isZoomedIn ? 'zoom in' : 'zoom out'), tabIndex: isClickable ? 0 : -1, "data-testid": 'image', isClickable: isClickable, isLoaded: isLoaded, isZoomedIn: isZoomedIn, isZoomable: isZoomable, wasImageClicked: wasImageClicked, className: placeholderClass || className, loading: 'lazy', onClick: handleImageClick, onKeyDown: handleImageClick, onLoad: function () { return setIsLoaded(true); } }, props), void 0)] }), void 0)) : (_jsxs(_Fragment, { children: [!isLoaded && (_jsx(StyledImagePlaceholder, { "data-testid": 'image-placeholder', className: placeholderClass || className }, void 0)),
            _jsx(StyledImage, __assign({ "data-testid": 'image', "aria-label": isClickable ? 'click me' : 'image', tabIndex: isClickable ? 0 : -1, className: className, isClickable: isClickable, isLoaded: isLoaded, isZoomedIn: isZoomedIn, isZoomable: isZoomable, wasImageClicked: wasImageClicked, loading: 'lazy', onClick: handleImageClick, onKeyDown: handleImageClick, onLoad: function () { return setIsLoaded(true); } }, props), void 0)] }, void 0));
};
Image.defaultProps = {
    id: '',
    className: '',
    style: {},
    placeholderClass: '',
    isClickable: false,
    isZoomable: false,
    src: '',
    alt: '',
    onClick: function () { return null; },
};
