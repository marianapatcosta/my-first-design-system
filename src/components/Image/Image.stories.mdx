import { Meta } from "@storybook/addon-docs/blocks";
import { Image } from "./Image";
import Test from '../../assets/images/test.jpg';

<Meta title="Design System|Image" component={Image} />

# Image
Simple Image component

## Usage
use image to render images and have a placeholder when the image is not loaded yet 
'Clickable' image can trigger a click event on click;
'Zoomable' image applies zoom in/out onclick; must have width defined;

- Default Image:
  <Story name='Default'>
    <Image src={Test}/>
  </Story>

- Clickable Image:
  <Story name='Clickable'>
    <Image src={Test} isClickable={true} onClick={() => console.log('clicked!')} />
  </Story>

- Zoomable Image:
  <Story name='Zoomable'>
    <div style={{ width: '300px' , height: '200px' }}>
      <Image src={Test} isZoomable={true} />
    </div>
  </Story>

## Other important notes
not applicable.