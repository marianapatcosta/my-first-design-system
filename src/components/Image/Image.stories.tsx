import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { Image, ImageProps } from './Image';
import Test from '../../assets/images/test.jpg';

export default {
  title: 'Image',
  component: Image,
  args: { src: Test, style: { width: '300px' } },
} as Meta;

const Template: Story<ImageProps> = (args) => <Image {...args} />;
const TemplateWrapped: Story<ImageProps> = (args) => (
  <div style={{ width: '300px', height: '200px' }}>
    <Image {...args} />
  </div>
);

export const Default = Template.bind({});
Default.args = {};

export const Clickable = Template.bind({});
Clickable.args = { isClickable: true, onClick: () => console.log('clicked!') };

export const Zoomable = TemplateWrapped.bind({});
Zoomable.args = { isZoomable: true };
