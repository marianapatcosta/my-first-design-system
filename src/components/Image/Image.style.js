var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
import styled, { keyframes, css } from 'styled-components';
import { baseStyles, colors } from '../../styles';
import { hexToRGB } from '../../utils/utils';
var scaleUp = keyframes(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  0% {\n    width: 50vw;\n  } \n  100% {\n    width: 95vw;\n  }\n"], ["\n  0% {\n    width: 50vw;\n  } \n  100% {\n    width: 95vw;\n  }\n"])));
var scaleDown = keyframes(templateObject_2 || (templateObject_2 = __makeTemplateObject(["\n   0% {\n    width: 100vw;\n  } \n  100% {\n    width: 100%;\n  }\n"], ["\n   0% {\n    width: 100vw;\n  } \n  100% {\n    width: 100%;\n  }\n"])));
export var StyledImageOverlay = styled.div(templateObject_3 || (templateObject_3 = __makeTemplateObject(["\n  ", "\n"], ["\n  ",
    "\n"])), function (_a) {
    var isZoomedIn = _a.isZoomedIn;
    return isZoomedIn &&
        "\n      height: 100vh;\n      width: 100vw;\n      position: fixed;\n      top: 0;\n      left: 0;\n      background-color: " + hexToRGB(colors.shadow, 0.9) + ";\n      display: flex;\n      justify-content: center;\n      align-items: center;\n      z-index: 100;\n    ";
});
export var StyledImage = styled.img(templateObject_6 || (templateObject_6 = __makeTemplateObject(["\n  ", "\n  width: 100%;\n  box-shadow: 0 0.125rem 0.5rem ", ";\n  cursor: default;\n\n  ", "\n\n  ", "\n  \n  ", "\n      \n  ", "\n"], ["\n  ", "\n  width: 100%;\n  box-shadow: 0 0.125rem 0.5rem ", ";\n  cursor: default;\n\n  ", "\n\n  ",
    "\n  \n  ",
    "\n      \n  ",
    "\n"])), baseStyles, colors.shadow, function (_a) {
    var isLoaded = _a.isLoaded;
    return !isLoaded && "visibility: hidden;";
}, function (_a) {
    var isZoomedIn = _a.isZoomedIn, isZoomable = _a.isZoomable;
    return isZoomable &&
        (isZoomedIn
            ? css(templateObject_4 || (templateObject_4 = __makeTemplateObject(["\n          animation: ", " 1s forwards;\n          -webkit-animation: ", " 1s forwards;\n          cursor: zoom-out;\n        "], ["\n          animation: ", " 1s forwards;\n          -webkit-animation: ", " 1s forwards;\n          cursor: zoom-out;\n        "])), scaleUp, scaleUp) : "cursor: zoom-in;");
}, function (_a) {
    var wasImageClicked = _a.wasImageClicked, isZoomedIn = _a.isZoomedIn;
    return wasImageClicked &&
        !isZoomedIn && css(templateObject_5 || (templateObject_5 = __makeTemplateObject(["\n      animation: ", " 0.3s forwards;\n      -webkit-animation: ", " 0.3s forwards;\n    "], ["\n      animation: ", " 0.3s forwards;\n      -webkit-animation: ", " 0.3s forwards;\n    "])), scaleDown, scaleDown);
}, function (_a) {
    var isClickable = _a.isClickable, isZoomable = _a.isZoomable;
    return isClickable && !isZoomable && "cursor: pointer;";
});
export var StyledImagePlaceholder = styled.div(templateObject_7 || (templateObject_7 = __makeTemplateObject(["\n  background-color: ", ";\n  position: relative;\n  overflow: hidden;\n  opacity: 0.5;\n  width: 100%;\n  height: 100%;\n  box-shadow: 0 0.125rem 0.5rem ", ";\n"], ["\n  background-color: ", ";\n  position: relative;\n  overflow: hidden;\n  opacity: 0.5;\n  width: 100%;\n  height: 100%;\n  box-shadow: 0 0.125rem 0.5rem ", ";\n"])), colors.secondary, colors.shadow);
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7;
