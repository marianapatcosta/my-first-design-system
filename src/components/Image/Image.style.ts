import styled, { keyframes, css } from 'styled-components';
import { baseStyles, colors } from '../../styles';
import { hexToRGB } from '../../utils/utils';

const scaleUp = keyframes`
  0% {
    width: 50vw;
  } 
  100% {
    width: 95vw;
  }
`;

const scaleDown = keyframes`
   0% {
    width: 100vw;
  } 
  100% {
    width: 100%;
  }
`;

export const StyledImageOverlay = styled.div<{
  isZoomedIn?: boolean;
}>`
  ${({ isZoomedIn }) =>
    isZoomedIn &&
    `
      height: 100vh;
      width: 100vw;
      position: fixed;
      top: 0;
      left: 0;
      background-color: ${hexToRGB(colors.shadow, 0.9)};
      display: flex;
      justify-content: center;
      align-items: center;
      z-index: 100;
    `}
`;

export const StyledImage = styled.img<{
  isClickable?: boolean;
  isLoaded?: boolean;
  isZoomedIn?: boolean;
  isZoomable?: boolean;
  wasImageClicked?: boolean;
}>`
  ${baseStyles}
  width: 100%;
  box-shadow: 0 0.125rem 0.5rem ${colors.shadow};
  cursor: default;

  ${({ isLoaded }) => !isLoaded && `visibility: hidden;`}

  ${({ isZoomedIn, isZoomable }) =>
    isZoomable &&
    (isZoomedIn
      ? css`
          animation: ${scaleUp} 1s forwards;
          -webkit-animation: ${scaleUp} 1s forwards;
          cursor: zoom-out;
        `
      : `cursor: zoom-in;`)}
  
  ${({ wasImageClicked, isZoomedIn }) =>
    wasImageClicked &&
    !isZoomedIn &&
    css`
      animation: ${scaleDown} 0.3s forwards;
      -webkit-animation: ${scaleDown} 0.3s forwards;
    `}
      
  ${({ isClickable, isZoomable }) =>
    isClickable && !isZoomable && `cursor: pointer;`}
`;

export const StyledImagePlaceholder = styled.div`
  background-color: ${colors.secondary};
  position: relative;
  overflow: hidden;
  opacity: 0.5;
  width: 100%;
  height: 100%;
  box-shadow: 0 0.125rem 0.5rem ${colors.shadow};
`;
