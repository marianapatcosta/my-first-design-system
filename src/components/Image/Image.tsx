import React, {
  CSSProperties,
  KeyboardEventHandler,
  MouseEventHandler,
  useEffect,
  useState,
} from 'react';
import { isEventValid } from '../../utils/utils';
import {
  StyledImageOverlay,
  StyledImage,
  StyledImagePlaceholder,
} from './Image.style';

export interface ImageProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
   * additional css class for image placeholder
   */
  placeholderClass?: string;
  /**
  /**
   * Image src attribute
   */
  src: string;
  /**
   * Image alt attribute
   */
  alt?: string;
  /**
   * is Image increasing size on click?
   */
  isZoomable?: boolean;
  /**
   * an event should be triggered on Image click? (Only available if isZoomable = false)
   */
  isClickable?: boolean;
  /**
   * Click handler (Only available if isZoomable = false)
   */
  onClick?:
    | MouseEventHandler<HTMLImageElement>
    | KeyboardEventHandler<HTMLImageElement>;
}

export const Image: React.FC<ImageProps> = ({
  placeholderClass,
  isClickable,
  isZoomable,
  onClick,
  ...props
}) => {
  const [isLoaded, setIsLoaded] = useState<boolean>(false);
  const [isZoomedIn, setIsZoomedIn] = useState<boolean>(false);
  const [wasImageClicked, setWasImageClicked] = useState<boolean>(false);

  const handleImageClick = (
    event: any /* MouseEvent<HTMLImageElement> | KeyboardEvent<HTMLImageElement> */
  ) => {
    if (!isEventValid(event) || (!isClickable && !isZoomable)) return;
    if (isZoomable) {
      setIsZoomedIn((prevIsZoomedIn) => !prevIsZoomedIn);
      !wasImageClicked && setWasImageClicked(true);
      return;
    }
    onClick && onClick(event);
  };

  useEffect(() => {
    if (isZoomedIn) {
      // document.getElementsByTagName('body')[0].style.overflow = 'hidden';
      document.documentElement.style.overflow = 'hidden';
      return;
    }

    // document.getElementsByTagName('body')[0].style.overflow = 'scroll';
    document.documentElement.style.overflow = 'scroll';
  }, [isZoomedIn]);

  return (
    <StyledImageOverlay isZoomable={isZoomable} isZoomedIn={isZoomedIn} >
      {!isLoaded && <StyledImagePlaceholder data-testid='image-placeholder' className={placeholderClass} />}
      <StyledImage
        data-testid='image'
        role={isClickable ? 'button' : 'img'}
        tabIndex={isClickable ? 0 : -1}
        isClickable={isClickable}
        isLoaded={isLoaded}
        isZoomedIn={isZoomedIn}
        isZoomable={isZoomable}
        wasImageClicked={wasImageClicked}
        loading='lazy'
        onClick={handleImageClick}
        onKeyDown={handleImageClick}
        onLoad={() => setIsLoaded(true)}
        {...props}
      />
    </StyledImageOverlay>
  );
};

Image.defaultProps = {
  id: '',
  className: '',
  style: {},
  placeholderClass: '',
  isClickable: false,
  isZoomable: false,
  src: '',
  alt: '',
  onClick: () => null,
};
