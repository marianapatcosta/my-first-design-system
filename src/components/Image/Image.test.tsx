import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Image } from './Image';
import Test from '../../assets/images/test.jpg';

describe('Image', () => {
  const defaultProps = { src: Test };
  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<Image {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should render image after onLoad call', () => {
    const { getByTestId } = render(<Image {...defaultProps} />);

    const img = getByTestId('image');
    const imgPlaceholder = getByTestId('image-placeholder');
    fireEvent.load(img);

    expect(img).toBeInTheDocument();
    expect(imgPlaceholder).not.toBeInTheDocument();
  });

  test('should not call onClick prop nor zoom image in on click if isZoomable prop nor isCLickable props are passed ', () => {
    const onClickMock = jest.fn();
   
    const { getByTestId } = render(
      <Image {...defaultProps}  />
    );

    const img = getByTestId('image');
    fireEvent.load(img);
    fireEvent.click(img);
    const style = window.getComputedStyle(img);
    expect(style.cursor).toBe('default');
    expect(onClickMock).not.toBeCalled();

  });

  test('should not call onClick prop on click if isZoomable prop is passed ', () => {
    const onClickMock = jest.fn();
    const { getByTestId } = render(
      <Image {...defaultProps} isZoomable={true} onClick={onClickMock}/>
    );

    const img = getByTestId('image');
    fireEvent.load(img);

    fireEvent.click(img);
  
    expect(onClickMock).not.toBeCalled();
  });

  test('should zoom image in on click if isZoomable prop is passed ', () => {
    const { getByTestId } = render(
      <Image {...defaultProps} isZoomable={true} />
    );

    const img = getByTestId('image');
    fireEvent.load(img);
    fireEvent.click(img);
    const style = window.getComputedStyle(img);
    expect(style.cursor).toBe('zoom-out');
  });

  test('should handle click events if isClickable prop is passed ', () => {
    const onClickMock = jest.fn();
    const { getByTestId } = render(
      <Image {...defaultProps} isClickable={true} onClick={onClickMock} />
    );
    const img = getByTestId('image');
    fireEvent.load(img);
    fireEvent.click(img);
    expect(onClickMock).toBeCalled();
  });

  test('should return null on click if isClickable prop is passed and onClick prop is not passed', () => {
    render(<Image {...defaultProps} isClickable={true} />);

    const result = Image.defaultProps?.onClick && Image.defaultProps.onClick();
    expect(result).toBeNull();
  });
});
