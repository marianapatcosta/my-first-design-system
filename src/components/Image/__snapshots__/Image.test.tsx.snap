// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders without errors and matches snapshot 1`] = `
<div
  class="sc-dlfnbm eUChx"
  data-testid="image-placeholder"
/>
`;
